File: frontend/lib/learn.ts
import api from "./api";

// ── Types ──
export type Concept = {
  concept: string;
  definition: string;
  importance: number;
  prerequisites: string[];
};

export type Flashcard = {
  id: string;
  concept: string | null;
  question: string;
  answer: string;
  difficulty: "easy" | "medium" | "hard";
  next_review_at: string;
};

export type QuizQuestion = {
  question: string;
  options: string[];
  correct_index: number;
  explanation: string;
};

export type LearnSession = {
  id: string;
  title: string;
  topic: string | null;
  source_type: "topic" | "text" | "file";
  subject: string | null;         // ← NEW
  summary: string | null;
  concepts: Concept[] | null;
  created_at: string;
};

export type LearnSessionDetail = LearnSession & {
  flashcards: Flashcard[];
  latest_quiz: {
    id: string;
    questions: QuizQuestion[];
    created_at: string;
  } | null;
};

export type StudyStats = {
  total_sessions: number;
  total_flashcards: number;
  due_today: number;
  mastered: number;
};

export type QuizResult = {
  score: number;
  correct: number;
  total: number;
  details: {
    question: string;
    your_answer: number | null;
    correct_index: number;
    explanation: string;
  }[];
};

// ── API calls ──
export const learnFromTopic = async (
  topic: string,
  numConcepts: number = 8,
  subject?: string,               // ← NEW
): Promise<LearnSessionDetail> => {
  const res = await api.post("/api/learn/from-topic", {
    topic,
    num_concepts: numConcepts,
    subject: subject ?? null,     // ← NEW
  });
  return res.data;
};

export const learnFromText = async (
  title: string,
  text: string,
  subject?: string,               // ← NEW
): Promise<LearnSessionDetail> => {
  const res = await api.post("/api/learn/from-text", {
    title, text,
    subject: subject ?? null,     // ← NEW
  });
  return res.data;
};

export const listLearnSessions = async (
  subject?: string,               // ← NEW
): Promise<LearnSession[]> => {
  const res = await api.get("/api/learn/sessions", {
    params: subject ? { subject } : {},
  });
  return res.data;
};

export const getLearnSession = async (
  id: string,
): Promise<LearnSessionDetail> => {
  const res = await api.get(`/api/learn/session/${id}`);
  return res.data;
};

export const deleteLearnSession = async (id: string): Promise<void> => {
  await api.delete(`/api/learn/session/${id}`);
};

export const reviewFlashcard = async (
  cardId: string,
  quality: number, // 0..5
): Promise<Flashcard> => {
  const res = await api.post(`/api/learn/flashcard/${cardId}/review`, {
    quality,
  });
  return res.data;
};

export const submitQuiz = async (
  attemptId: string,
  answers: number[],
): Promise<QuizResult> => {
  const res = await api.post(`/api/learn/quiz/${attemptId}/submit`, {
    answers,
  });
  return res.data;
};

export const getLearnStats = async (): Promise<StudyStats> => {
  const res = await api.get("/api/learn/stats");
  return res.data;
};

export const learnFromFile = async (
  file: File,
  title?: string,
  subject?: string,               // ← NEW
): Promise<LearnSessionDetail> => {
  const formData = new FormData();
  formData.append("file", file);
  if (title) formData.append("title", title);
  if (subject) formData.append("subject", subject);    // ← NEW
  const res = await api.post("/api/learn/from-file", formData, {
    headers: { "Content-Type": "multipart/form-data" },
    timeout: 120000,
  });
  return res.data;
};

// ── Mind Map ──
export type MindMapNode = {
  id: string;
  label: string;
  description: string;
  level: number;
  parent: string | null;
};

export type MindMapEdge = {
  id: string;
  source: string;
  target: string;
  label: string | null;
};

export type MindMapGraph = {
  nodes: MindMapNode[];
  edges: MindMapEdge[];
};

export const getSessionMindMap = async (
  sessionId: string,
): Promise<MindMapGraph> => {
  const res = await api.get(`/api/learn/session/${sessionId}/mindmap`, {
    timeout: 60000,
  });
  return res.data;
};

File: backend/app/schemas/learn.py
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


# ── Requests ──
class LearnFromTopicRequest(BaseModel):
    topic: str = Field(min_length=2, max_length=200)
    num_concepts: int = Field(default=8, ge=4, le=12)
    subject: str | None = None   
    class_name: str | None = None    


class LearnFromTextRequest(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    text: str = Field(min_length=20)
    subject: str | None = None        # ← NEW


class ReviewFlashcardRequest(BaseModel):
    quality: int = Field(ge=0, le=5)


class SubmitQuizRequest(BaseModel):
    answers: list[int]


# ── Responses ──
class ConceptOut(BaseModel):
    concept: str
    definition: str
    importance: int = 5
    prerequisites: list[str] = []


class FlashcardOut(BaseModel):
    id: UUID
    concept: str | None
    question: str
    answer: str
    difficulty: str
    next_review_at: datetime

    class Config:
        from_attributes = True


class QuizQuestionOut(BaseModel):
    question: str
    options: list[str]
    correct_index: int
    explanation: str


class SessionOut(BaseModel):
    id: UUID
    title: str
    topic: str | None
    source_type: str
    subject: str | None = None        # ← NEW
    summary: str | None
    concepts: list[dict[str, Any]] | None
    created_at: datetime

    class Config:
        from_attributes = True


class SessionDetailOut(SessionOut):
    flashcards: list[FlashcardOut]
    latest_quiz: dict[str, Any] | None = None


class StudyStatsOut(BaseModel):
    total_sessions: int
    total_flashcards: int
    due_today: int
    mastered: int